examples/hackweek_demo/plot_pnl.py

```diff
@@ -1,204 +1,206 @@
+        for row in self.deltas.index[1:]:
+            # An update is required for columns which increase in size this row, identified by a positive delta.
+            update_required = self.deltas.loc[row, :] > 0
+
+            new_price = []
+            if len(update_required) > 0:
+                # calculate update, per this general formula:
+                # new_price = (delta_amount * current_price + old_amount * old_price) / (old_amount + delta_amount)
+                new_price = (
+                    share_price_on_increases.loc[row, update_required] * self.deltas.loc[row, update_required]
+                    + self.open_share_price.loc[row - 1, update_required] * self.positions.loc[row - 1, update_required]
+                ) / (self.deltas.loc[row, update_required] + self.positions.loc[row - 1, update_required])
+
+            # Keep previous result where an update isn't required, otherwise replace with new_price
+            self.open_share_price.loc[row, :] = self.open_share_price.loc[row - 1, :].where(
+                ~update_required, new_price, axis=0
             )
-            # print(f"{spot_price=}")
 
-            # LP value (TODO: check why certain agents own more than 100% of the pool)
-            total_lp_value = row.share_reserves / 1e18 * row.share_price / 1e18 + row.bond_reserves / 1e18 * spot_price
-            share_of_pool = float(agent_wallets[agent].lp_tokens) / (row.lp_total_supply / 1e18)
-            # print(f"{float(agent_wallets[agent].lp_tokens)=}")
-            # print(f"{row.lp_total_supply/1e18=}")
-            pnl = share_of_pool * total_lp_value
 
-            # for each LONG
-            for _, long in agent_wallets[agent].longs.items():
-                pnl += float(long.balance) * spot_price
+def get_agent_positions(session: Session) -> dict[str, AgentPosition]:
+    """Create an AgentPosition for each agent in the wallet history."""
+    return {agent: AgentPosition(wallet) for agent, wallet in pg.get_wallet_info_history(session).items()}
 
-            # for each SHORT
-            for _, short in agent_wallets[agent].shorts.items():
-                pnl += float(short.balance) * (1 - spot_price)
 
-            new_pnl[f"{agent_shortname}"] = pnl
-        pnl_data.append(new_pnl)
-        time_data.append(row["timestamp"])
-
-    # %%
-    y_data = pd.DataFrame(pnl_data)
-    x_data = pd.DataFrame(time_data)
-    return (x_data, y_data)
+def calculate_pnl(
+    pool_config: pd.DataFrame, pool_info: pd.DataFrame, checkpoint_info: pd.DataFrame
+) -> dict[str, AgentPosition]:
+    """Calculate pnl for all agents.
 
+    Arguments
+    ---------
+    pool_config : PoolConfig
+        Configuration with which the pool was initialized.
+    pool_info : pd.DataFrame
+        Reserves of the pool at each block.
+    checkpoint_info : pd.DataFrame
+    Checkpoint information at each block.
+    """
+    session = pg.initialize_session()
```

Comment:
this should be passed in.  even better, agent_positions should be passed in.  We should talk about this to clarify but  very few functions should know about postgres.  There should be specific functions that go get a thing from postgres.  If a function has 'calculate' in the name that's a signal that its just doing math things, it shouldn't fetch data.

Reply:
moved it to postgres, cus it's really a data manipulation task, and matches this pattern in run_demo.py:
```py
    pool_config_data = postgres.get_pool_config(session)
    txn_data = postgres.get_transactions(session, start_block=start_block)
    pool_info_data = postgres.get_pool_info(session, start_block=start_block)
    checkpoint_info = postgres.get_checkpoint_info(session, start_block=start_block)
    agent_positions = postgres.get_agent_positions(session)
```